modules/file-transfer/src/main/java/org/apache/ignite/internal/network/file/FileSender.java

```diff
@@ -0,0 +1,227 @@
+
+import java.io.IOException;
+import java.nio.file.Path;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Queue;
+import java.util.UUID;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Semaphore;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.stream.Collectors;
+import org.apache.ignite.internal.network.file.exception.FileTransferException;
+import org.apache.ignite.network.MessagingService;
+
+/**
+ * A class that sends files to a node. It uses a rate limiter to limit the bandwidth used. It also uses a thread pool to send the files in
+ * parallel.
+ */
+class FileSender {
+    private final int chunkSize;
+
+    private final Semaphore rateLimiter;
+
+    private final MessagingService messagingService;
+
+    private final ExecutorService executorService;
+
+    private final Queue<FileTransfer> requests = new LinkedList<>();
+
+    FileSender(
+            int chunkSize,
+            Semaphore rateLimiter,
+            MessagingService messagingService,
+            ExecutorService executorService
+    ) {
+        this.chunkSize = chunkSize;
+        this.rateLimiter = rateLimiter;
+        this.messagingService = messagingService;
+        this.executorService = executorService;
+    }
+
+    /**
+     * Creates {@link FileTransfer} objects for each file and places them in the queue. The files will be sent when the rate limiter is
+     * available.
+     *
+     * @param targetNodeConsistentId The consistent ID of the node to send the files to.
+     * @param transferId The ID of the transfer.
+     * @param paths The paths of the files to send.
+     * @return A future that will be completed when the transfer is complete.
+     */
+    CompletableFuture<Void> send(String targetNodeConsistentId, UUID transferId, List<Path> paths) {
+        // It doesn't make sense to continue file transfer if there is a failure in one of the files.
+        AtomicBoolean shouldBeCancelled = new AtomicBoolean(false);
+
+        List<FileTransfer> transfers = paths.stream()
+                .map(path -> new FileTransfer(targetNodeConsistentId, transferId, path, shouldBeCancelled))
+                .collect(Collectors.toList());
+
+        CompletableFuture[] results = transfers.stream().map(it -> it.result)
+                .map(it -> it.whenComplete((v, e) -> {
+                    if (e != null) {
+                        shouldBeCancelled.set(true);
+                    }
+                }))
+                .toArray(CompletableFuture[]::new);
+
+        transfers.forEach(this::processTransfer);
+
+        return allOf(results);
+    }
+
+    /**
+     * Processes the given transfer. If the rate limiter is not available, the transfer will be processed later.
+     *
+     * @param transfer The transfer to process.
+     */
+    private void processTransfer(FileTransfer transfer) {
+        synchronized (rateLimiter) {
```

Comment:
Please document, how and why you use this synchronization. If it protects the "requests" queue, then why do you use rate limiter as a monitor? Not very intuitive to me